### PR TITLE
[FEATURE REQUEST] List links over a space

### PR DESCRIPTION
**Pull request by joragua**
## Related Issues
App: https://github.com/owncloud/android/issues/4752

- [x] Add changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
- [x] Add feature to Release Notes in `ReleaseNotesViewModel.kt` creating a new `ReleaseNote()` with String resources (if required)

_____

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/Android/Executions/Release_4.8/Spaces%20-%20Show%20Links.md

Reports:

- [x] (1) Chain icon in talkback as `button` https://github.com/owncloud/android/pull/4783#issuecomment-3996382256 [FIXED]
- [x] (2) Scroll over links list only works if no links visible https://github.com/owncloud/android/pull/4783#issuecomment-3996955254 [FIXED]
- [x] (3) Spinner while lists are loading https://github.com/owncloud/android/pull/4783#issuecomment-3996965509 [DONE]

### CODE REVIEW

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        val view = inflater.inflate(R.layout.public_link_item, parent, false)
+        view.filterTouchesWhenObscured = PreferenceUtils.shouldDisallowTouchesWithOtherVisibleWindows(parent.context)
+
+        return SpaceLinksViewHolder(view)
+    }
+
+    override fun onBindViewHolder(holder: SpaceLinksViewHolder, position: Int) {
+        val spaceLink = spaceLinks[position]
+        holder.binding.apply {
+            publicLinkDisplayName.text = spaceLink.displayName
```
Comment (jesmrec) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt:54]: `displayName` is not mandatory, did you take that in account for the view?

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            publicLinkType.text = holder.itemView.context.getString(spaceLink.type.toStringResId())
+
+            val hasExpirationDate = spaceLink.expirationDateTime != null
+            expirationCalendarIcon.isVisible = hasExpirationDate
+            expirationDate.isVisible = hasExpirationDate
+            if (hasExpirationDate) {
+                expirationDate.apply {
+                    text = DisplayUtils.displayDateToHumanReadable(spaceLink.expirationDateTime)
+                    contentDescription = holder.itemView.context.getString(R.string.content_description_member_expiration_date, expirationDate.text)
```
Comment (jesmrec) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt:63]: does `context` need to be referenced with `holder.itemView` ? Couldn't be used `context` instead (with no `holder.itemView` ?

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersActivity.kt [hunk 3]
```diff
@@ -100,18 +100,22 @@ class SpaceMembersActivity: FileActivity(), SpaceMembersFragment.SpaceMemberFrag
...
     }
 
-    private fun copyOrSendPermanentLink(permanentLink: String?, spaceName: String) {
-        permanentLink?.let {
+    override fun copyOrSendPublicLink(publicLinkUrl: String, spaceName: String) {
+        copyOrSendLink(publicLinkUrl, spaceName)
+    }
+
+    private fun copyOrSendLink(link: String?, spaceName: String) {
+        link?.let {
```
Comment (jesmrec) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersActivity.kt:108]: can the link be `null` somehow?

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt [hunk 4]
```diff
@@ -251,8 +268,25 @@ class SpaceMembersFragment : Fragment(), SpaceMembersAdapter.SpaceMembersAdapter
...
+            noPublicLinksMessage.isVisible = !hasLinks
+        }
+    }
+
+    private fun showSpaceLinks(spaceLinks: List<OCLink>) {
+        val formatter = SimpleDateFormat(DisplayUtils.DATE_FORMAT_ISO, Locale.ROOT).apply {
+            timeZone = TimeZone.getTimeZone("UTC")
+        }
+        spaceLinksAdapter.setSpaceLinks(spaceLinks.sortedByDescending { spaceLink ->
+            formatter.parse(spaceLink.createdDateTime)
```
Comment (jesmrec) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt]: Are we totally sure `createdDateTime` will be always there? i guess it will. Anyway, should/could it be protected?

File: owncloudApp/src/main/res/layout/members_fragment.xml [hunk 5]
```diff
@@ -70,12 +70,64 @@
...
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:paddingTop="@dimen/standard_padding"
+            android:paddingBottom="@dimen/standard_padding"
+            android:paddingStart="@dimen/standard_half_padding"
+            android:paddingEnd="@dimen/standard_half_padding"
+            android:text="@string/share_no_public_links"
+            android:textSize="15sp"
+            android:visibility="gone"/>
+
```
Comment (jesmrec) [inline owncloudApp/src/main/res/layout/members_fragment.xml:138]: One question about this file structure:
```
<ConstraintLayout>
  <RecyclerView>
</ConstraintLayout>
<ConstraintLayout>
  <RecyclerView>
</ConstraintLayout>
```

inside of a `<LinearLayout>`. 

Would it be more maintainable to separate every 

```
<ConstraintLayout>
  <RecyclerView>
</ConstraintLayout>
```
in different files (one for members, other for links), that are referenced with `include` in `members_fragment` ??

File: owncloudApp/src/main/res/layout/public_link_item.xml [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            android:visibility="visible"/>
+
+    </LinearLayout>
+
+    <View
+        android:layout_width="match_parent"
+        android:layout_height="1dp"
+        android:layout_marginTop="@dimen/standard_quarter_margin"
+        android:background="@android:color/darker_gray"
+        android:alpha="0.5"/>
```
Comment (jesmrec) [inline owncloudApp/src/main/res/layout/public_link_item.xml:119]: What is this `View` for?

File: owncloudApp/src/main/res/layout/public_link_item.xml [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+</LinearLayout>
```
Comment (jesmrec) [inline owncloudApp/src/main/res/layout/public_link_item.xml:121]: Is accessibility completed in this layout?

File: owncloudDomain/src/main/java/com/owncloud/android/domain/links/model/OCLink.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    val displayName: String,
+    val type: OCLinkType,
+    val webUrl: String
+) : Parcelable
+
+enum class OCLinkType {
+    CAN_VIEW, CAN_EDIT, CAN_UPLOAD, CREATE_ONLY, INTERNAL;
+
+    companion object {
+
```
Comment (jesmrec) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/links/model/OCLink.kt]: blank line to remove?

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/spaces/responses/SpacesResponse.kt [hunk 9]
```diff
@@ -104,15 +104,15 @@ data class SpecialFolderResponse(
...
     val expirationDateTime: String?,
+
+    // Member response
     val grantedToV2: GrantedToV2Response?,
-    val id: String?,
     val roles: List<String>?,
 
     // Link response
-    val createDateTime: String?,
-    val hasPassword: Boolean?,
```
Comment (jesmrec) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/spaces/responses/SpacesResponse.kt:115]: why removing the `hasPassword`? it's correct that it is not relevant for the current PR, but links have passwords and could be useful for the following ones.

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt (hunk 1 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt:54]: Yes, but it's a corner case I think 🤔 In the clients, it's not possible to create a link without name... Anyway, it should show an empty string (`""`) if the `displayName` does not exist

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt (hunk 2 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt:63]: Yes! In adapters, the `context` must be referenced with `holder.itemView.context`

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersActivity.kt (hunk 3 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersActivity.kt:108]: Following `OCSpace` model, `webUrl` property (used for permalink) can be null. That's the reason I added `link?.let` in the function

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt (hunk 4 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt]: It should... But I don't know if there could be some corner cases 🤔 In all tests that I did,  all links had `createdDateTime` property...

File: owncloudApp/src/main/res/layout/public_link_item.xml (hunk 6 above)
Comment (joragua) [reply owncloudApp/src/main/res/layout/public_link_item.xml:119]: It's the gray divider below each link (the same one used for space members) 😄

File: owncloudApp/src/main/res/layout/public_link_item.xml (hunk 7 above)
Comment (joragua) [reply owncloudApp/src/main/res/layout/public_link_item.xml:121]: It should. There are more accessibility properties (in `SpaceMembersAdapter`) which are set dynamically

File: owncloudDomain/src/main/java/com/owncloud/android/domain/links/model/OCLink.kt (hunk 8 above)
Comment (joragua) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/links/model/OCLink.kt]: Removed! ✅

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/spaces/responses/SpacesResponse.kt (hunk 9 above)
Comment (joragua) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/spaces/responses/SpacesResponse.kt:115]: Good catch! I will keep `hasPassword` property for future issues 🙌🏻 It can be interesting and useful for create/edit link issues

File: owncloudApp/src/main/res/layout/members_fragment.xml (hunk 5 above)
Comment (joragua) [reply owncloudApp/src/main/res/layout/members_fragment.xml:138]: It would be interesting if they are used in other XML files but this is the only use for these layouts. What do you think? 🤔 On the other hand, all `RecyclerView` are not inside `ConstraintLayout`. The file structure of `members_fragment.xml` is:

```
<LinearLayout>
   <ConstraintLayout> (members header)
   <RecyclerView> (members list)
   <ConstraintLayout> (public links header)
   <TextView> (warning header)
   <RecyclerView> (public links list)
</LinearLayout>
```

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt (hunk 1 above)
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt:54]: Will take a look in QA just in case

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt (hunk 2 above)
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/links/SpaceLinksAdapter.kt:63]: 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersActivity.kt (hunk 3 above)
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersActivity.kt:108]: Ok, it's better to be cautious, even when the link will be never `null`

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt (hunk 4 above)
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt]: What would happen if `createdDateTime` would be `null`?

File: owncloudApp/src/main/res/layout/members_fragment.xml (hunk 5 above)
Comment (jesmrec) [reply owncloudApp/src/main/res/layout/members_fragment.xml:138]: ok, makes sense.

File: owncloudDomain/src/main/java/com/owncloud/android/domain/links/model/OCLink.kt (hunk 8 above)
Comment (jesmrec) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/links/model/OCLink.kt]: 👍

File: owncloudApp/src/main/res/layout/public_link_item.xml (hunk 7 above)
Comment (jesmrec) [reply owncloudApp/src/main/res/layout/public_link_item.xml:121]: Will check in QA phase

File: owncloudApp/src/main/res/layout/public_link_item.xml (hunk 6 above)
Comment (jesmrec) [reply owncloudApp/src/main/res/layout/public_link_item.xml:119]: Oh, thats right! reading the layout , i did not realised. Ok!

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/spaces/responses/SpacesResponse.kt (hunk 9 above)
Comment (jesmrec) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/spaces/responses/SpacesResponse.kt:115]: Password back!

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt (hunk 4 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt]: Hmm.. It is parsed to an empty string (`""`) but the app it will crash because is not a valid string to format...

I've added a new condition: `Date(0)` (minimal date) is set and compared with other dates when the `createdDateTime` does not exist. All public links with no creation date will be displayed at the end of the list. What do you think about this solution?

Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/members/SpaceMembersFragment.kt]: Cool! this scenario should not happen, but, if for any case it'd happen, the app is safe.

**Comment by jesmrec** [comment]
Moving to QA

**Comment by mmattel** [comment]
Hooking myself in just to monitor upcoming changes in Transifex

**Comment by jesmrec** [comment]
### (1) [FIXED]

About talkback: when clicking on the chain icon in the link cell, it just says "Button". Not very descriptive, something like "Copy link" or similar would be better. Also, including the name of the link would be the best approach

Xiaomi Redmi 13 Note, Android 15
`cab9c66e1`

**Comment by joragua** [comment]
**(1)** should be fixed with the commit 2b401c7cf6cac143e4249236c75a6102b4ca4a70

**Comment by jesmrec** [comment]
(1) fixed

**Comment by jesmrec** [comment]
### (2) [FIXED]

Take a setup like this one:

Screenshot 2026-03-04 at 12 29 14

with a list of members that take almost all the screen (not all), and the first link of the list below. In this case, the global scroll does not work. In order to scroll the list of links, an accurated scroll must be done over the unique displayed cell (bad in UX terms)

In case there is no links visible, the global scroll works fine.

Xiaomi Redmi 13 Note, Android 15
Samsung Galaxy Tab A8, Android 15
`2b401c7cf6`

**Comment by jesmrec** [comment]
### (3) (improvement) [DONE]

In there is a long list of members and links, a short bit of time is take to load the list. It could be covered with an spinner to let users know that device is working and not frozen 

Xiaomi Redmi 13 Note, Android 15
Samsung Galaxy Tab A8, Android 15
`2b401c7cf6`

**Comment by jesmrec** [comment]
Approved on my side! 🚀
